src/concurrency/genmc/scheduling.rs

```diff
@@ -1,58 +1,120 @@
 use genmc_sys::{ActionKind, ExecutionState};
+use rustc_middle::mir::TerminatorKind;
+use rustc_middle::ty::{self, Ty};
 
 use super::GenmcCtx;
 use crate::{
     InterpCx, InterpResult, MiriMachine, TerminationInfo, ThreadId, interp_ok, throw_machine_stop,
 };
 
+enum NextInstrKind {
+    MaybeAtomic(ActionKind),
+    NonAtomic,
+}
+
+/// Check if a call or tail-call could have atomic load semantics.
+fn get_next_instruction_kind<'tcx>(
```

Comment:
This improved logic doesn't seem to affect any tests. Shouldn't this reduce the number of executions because we now schedule better?